### PR TITLE
Remove aiofiles dependency

### PR DESCRIPTION
**Pull request by b-rowan**
Remove dependency on aiofiles, `NamedTemporaryFile` was causing issues. Replace temporary files with a temporary file in `artifacts/tmp`.

### CODE REVIEW

File: goosebit/api/v1/software/routes.py [hunk 1]
```diff
@@ -69,11 +71,15 @@ async def post_update(_: Request, file: UploadFile | None = File(None), url: str
...
 
-        async with aiofiles.tempfile.NamedTemporaryFile("w+b") as f:
+        async with await open_file(tmp_file_path, "w+b") as f:
             await f.write(await file.read())
-            absolute = await file_path.absolute()
-            software = await create_software_update(absolute.as_uri(), Path(f.name))
+        absolute = await file_path.absolute()
+        tmp_absolute = await tmp_file_path.absolute()
+        software = await create_software_update(absolute.as_uri(), tmp_absolute)
+        await tmp_file_path.unlink(missing_ok=True)
```
Comment (tsagadar) [inline goosebit/api/v1/software/routes.py:82]: Move temp file deletion in a finally block to ensure that it always gets deleted.

File: goosebit/updates/swdesc.py [hunk 2]
```diff
@@ -68,9 +71,14 @@ async def parse_file(file: Path):
...
         file = await c.get(url)
-        async with aiofiles.tempfile.NamedTemporaryFile("w+b") as f:
+        artifacts_dir = Path(config.artifacts_dir)
+        tmp_file_path = artifacts_dir.joinpath("tmp", ("".join(random.choices(string.ascii_lowercase, k=12)) + ".tmp"))
+        await tmp_file_path.parent.mkdir(parents=True, exist_ok=True)
+        async with await open_file(tmp_file_path, "w+b") as f:
             await f.write(file.content)
-            return await parse_file(Path(str(f.name)))
+        file_data = await parse_file(Path(str(f.name)))
+        await tmp_file_path.unlink()
```
Comment (tsagadar) [inline goosebit/updates/swdesc.py:80]: Move temp file deletion in a finally block to ensure that it always gets deleted.

File: goosebit/updates/__init__.py [hunk 3]
```diff
@@ -19,7 +19,9 @@ async def create_software_update(uri: str, temp_file: Path | None) -> Software:
     parsed_uri = urlparse(uri)
 
     # parse swu header into update_info
-    if parsed_uri.scheme == "file" and temp_file is not None:
+    if parsed_uri.scheme == "file":
+        if temp_file is None:
+            raise HTTPException(422, "Temporary file missing, cannot parse file information")
```
Comment (tsagadar) [inline goosebit/updates/__init__.py:24]: I think a better error code would be 500. This is an internal server issue - nothing caused by the client request.

**Review by tsagadar — CHANGES_REQUESTED** [review]

**Comment by easybe** [comment]
> Remove dependency on aiofiles, `NamedTemporaryFile` was causing issues. Replace temporary files with a temporary file in `artifacts/tmp`.

What kind of issues?

File: goosebit/api/v1/software/routes.py [hunk 4]
```diff
@@ -69,11 +71,15 @@ async def post_update(_: Request, file: UploadFile | None = File(None), url: str
...
         file_path = artifacts_dir.joinpath(file.filename)
+        tmp_file_path = artifacts_dir.joinpath("tmp", ("".join(random.choices(string.ascii_lowercase, k=12)) + ".tmp"))
+        await tmp_file_path.parent.mkdir(parents=True, exist_ok=True)
 
-        async with aiofiles.tempfile.NamedTemporaryFile("w+b") as f:
+        async with await open_file(tmp_file_path, "w+b") as f:
             await f.write(await file.read())
-            absolute = await file_path.absolute()
-            software = await create_software_update(absolute.as_uri(), Path(f.name))
+        absolute = await file_path.absolute()
```
Comment (easybe) [inline goosebit/api/v1/software/routes.py:79]: I don’t find `absolute` a good variable name. It doesn’t suggest what it is storing. Absolut Vodka? 😉

**Comment by b-rowan** [comment]
> > Remove dependency on aiofiles, `NamedTemporaryFile` was causing issues. Replace temporary files with a temporary file in `artifacts/tmp`.
> 
> 
> 
> What kind of issues?

Was running into permissions errors.  It wasn't entirely aiofiles, but the way we were opening and closing files.  We opened the temp file at the top level, then tried to open it again in a function called inside that with block, and that was causing issues.

In this case though, I don't really see any reason to use aiofiles or temp files here, there doesn't seem like much benefit...

File: goosebit/api/v1/software/routes.py (hunk 4 above)
Comment (b-rowan) [reply goosebit/api/v1/software/routes.py:79]: That would be a heck of a variable 😂

I'll get this renamed.

**Comment by tsagadar** [comment]
LGTM. Great to have that dependency removed!

**Review by tsagadar — APPROVED** [review]

File: goosebit/api/v1/software/routes.py [hunk 5]
```diff
@@ -69,11 +71,16 @@ async def post_update(_: Request, file: UploadFile | None = File(None), url: str
         # local file
         artifacts_dir = Path(config.artifacts_dir)
         file_path = artifacts_dir.joinpath(file.filename)
-
-        async with aiofiles.tempfile.NamedTemporaryFile("w+b") as f:
-            await f.write(await file.read())
-            absolute = await file_path.absolute()
-            software = await create_software_update(absolute.as_uri(), Path(f.name))
+        tmp_file_path = artifacts_dir.joinpath("tmp", ("".join(random.choices(string.ascii_lowercase, k=12)) + ".tmp"))
```
Comment (easybe) [inline goosebit/api/v1/software/routes.py:74]: This change makes the artifact directory mandatory. Before, it was optional (when only using external artifacts) as temporary files were created in `/tmp` :disappointed:

Comment (b-rowan) [reply goosebit/api/v1/software/routes.py:74]: Didn't even think about this.  Maybe I can generate a temporary file location other than this?

Comment (easybe) [reply goosebit/api/v1/software/routes.py:74]: IMHO it is best to use a standard temp location. But, we should make sure it works on all OSes.

**Pull request merged**